examples/svdiff/modeling_svdiff.py

```diff
@@ -0,0 +1,188 @@
+import torch
+from torch import nn
+from torch.nn import functional as F
+from einops import rearrange
+import inspect
+import safetensors
+from accelerate.utils import set_module_tensor_to_device
+
+
+class SVDiffModule(nn.Module):
+    def __init__(self, weight: torch.Tensor, weight_type: Optional[str] = None, init_delta: Optional[torch.Tensor] = None, scale: float = 1.0):
+        """
+        SVDiff module to be registered as hook 
+
+        Parameters:
+            weight (`torch.Tensor`, *required*): pre-trained weight 
+            weight_type (`string`, *optional*): choose from [None, 'conv', '1d']
+            init_delta (`string, *optional*`): init values for delta
+            scale (`float`): spectral shifts scale. This is intended to use during inference. 
+        """
+        super().__init__()
+        self.weight_type = weight_type
+        weight = self._reshape_weight_for_svd(weight)
+        # perform SVD
+        with torch.no_grad():
+            U, S, Vh = torch.linalg.svd(weight, full_matrices=False)
+        self.register_buffer("U", U.detach())
+        self.U.requires_grad = False
+        self.register_buffer("S", S.detach())
+        self.S.requires_grad = False
+        self.register_buffer("Vh", Vh.detach())
+        self.Vh.requires_grad = False
+        # initialize to 0 for smooth tuning 
+        self.delta = nn.Parameter(torch.zeros_like(S)) if init_delta is None else nn.Parameter(init_delta)
+        self.scale = scale
+
+    def _reshape_weight_for_svd(self, updated_weight: torch.Tensor, original_weight: torch.Tensor = None, reverse: bool = False, **kwargs):
+        if self.weight_type is not None:
+            if self.weight_type == "conv":
+                if not reverse:
+                    return rearrange(updated_weight, 'co cin h w -> co (cin h w)')
+                else:
+                    return rearrange(updated_weight, 'co (cin h w) -> co cin h w', cin=original_weight.size(1), h=original_weight.size(2), w=original_weight.size(3))
+            elif self.weight_type == "1d":
+                if not reverse:
+                    return updated_weight.unsqueeze(0)
+                else:
+                    return updated_weight.squeeze(0)
+            else:
+                raise ValueError(f"`weight_type`={self.weight_type} is invalid!")
+        else:
+            return updated_weight
+        
+    def forward(self, weight: torch.Tensor):
+        updated_weight = self.U @ torch.diag(F.relu(self.S + self.scale * self.delta)) @ self.Vh
+        updated_weight = self._reshape_weight_for_svd(updated_weight, weight, reverse=True)
+        return updated_weight
+
+
+def set_spectral_shifts(model: nn.Module, spectral_shifts_ckpt: str = None, **kwargs):
+    # key to module
+    svdiff_modules = {}
+    for name, module in model.named_modules():
+        svdiff_module = None
+        # TODO: currently only support classes in SD. Needs to be more general (eg. there are more classes for 1-D weights)
```

Comment:
Could you clarify what this is about?

Reply:
I think it's okay to remove. 
SVD can be performed in any weights defined in pytorch theoretically. But, I only confirmed modules inside SD (`nn.Conv1D`, `nn.Linear` etc.). So, the other modules might be added in the future.

Reply:
https://github.com/huggingface/diffusers/pull/3107/commits/f9f8bc0292021e52337871bf179971c8aed5c3f3